slatedb/src/cached_object_store/object_store.rs

```diff
@@ -92,15 +151,51 @@ impl CachedObjectStore {
         })
     }
 
-    // TODO: implement the put with cache here
-    #[allow(unused)]
     async fn cached_put_opts(
         &self,
         location: &Path,
         payload: object_store::PutPayload,
         opts: object_store::PutOptions,
     ) -> object_store::Result<PutResult> {
-        self.object_store.put_opts(location, payload, opts).await
+        // Only cache if the cache_puts option is enabled
+        if !self.cache_puts {
+            // If caching is disabled, just write to the upstream object store without cloning
+            return self.object_store.put_opts(location, payload, opts).await;
+        }
+
+        // First, write to the upstream object store (cloning payload is cheap since it's just a Arc internally)
+        let result = self
+            .object_store
+            .put_opts(location, payload.clone(), opts)
+            .await?;
+
+        // Then, save to local cache if admission policy allows it
+        // Get metadata and attributes using our cached head method (this also saves the head to cache)
+        match self.cached_head(location).await {
```

Comment:
We should keep an eye on perf since this adds a second round trip to the object store. It might be worth doing this as a separate async spawned task to unblock the put call.

Reply:
I'm looking the get code path in `maybe_prefetch_range` that is has to do `read_head` from the entry 
https://github.com/slatedb/slatedb/blob/main/slatedb/src/cached_object_store/object_store.rs#L118 

That's why I thought it's required in the read path.

Reply:
Agh, yea, that's what I was worried about.

@flaneur2020 WDYT about changing maybe_prefetch_range to do a check against the current cache and split the range up so we use existing partitions if they're available (rather than assuming they are/aren't based on header)?

Reply:
@nomiero Let's keep things as they are in the current PR.